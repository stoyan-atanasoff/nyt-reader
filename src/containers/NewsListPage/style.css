div.loader {
  text-align: center;
  margin-top: 50px;
}
div.news-message {
  margin-top: 50px;
}
