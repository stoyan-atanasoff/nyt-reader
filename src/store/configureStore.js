import { createBrowserHistory } from 'history';
import { createStore, applyMiddleware, compose } from 'redux';
import createSagaMiddleware from 'redux-saga';
import sagas from '../sagas';
import createReducer from '../reducers';
import {routerMiddleware} from "connected-react-router";

export const history = createBrowserHistory();

export default function (initialState = {}) {
  let composeEnhancers = compose;
  const reduxSagaMonitorOptions = {};

  // if (typeof window === 'object') {
  //   if (window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__)
  //     composeEnhancers = window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__({});
  // }
  const sagaMiddleware = createSagaMiddleware(reduxSagaMonitorOptions);
  const middleWares = [sagaMiddleware, routerMiddleware(history)];
  const enhancers = [applyMiddleware(...middleWares)];
  const store = createStore(createReducer(), initialState, composeEnhancers(...enhancers));
  sagaMiddleware.run(sagas);

  return store;
}
