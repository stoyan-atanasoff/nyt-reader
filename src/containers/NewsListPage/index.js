import React, { memo } from 'react';
import {compose} from "redux";
import {connect} from "react-redux";
import {Alert, Col, Row, Spin} from "antd";
import {fetchNews} from "../../actions/newsActions";
import {selectNewsList, selectNewsLoading, selectNewsMessage, selectNewsStatus} from "../../selectors/newsSelectors";
import NewsListItem from "../../components/NewsListItem";
import './style.css';
import PropTypes from "prop-types";

class NewsListPage extends React.Component {
  componentDidMount() {
    if (!this.props.newsList.length) {
      this.props.fetchNews();
    }
  }

  render() {
    // Show loading bar
    if (this.props.newsLoading) return <div className='loader'><Spin size='default' /><br />{this.props.newsMessage}</div>;

    // Show error message
    if (!this.props.newsStatus) return <Row type='flex' lign='top' justify='center' className={'news-message'}><Col><Alert message={this.props.newsMessage} type="error" /></Col></Row>;

    // Show news
    let cards = 1;
    return <Row type='flex' lign='top'>
      {this.props.newsList.map(news => <Col xs={24} sm={24} key={news.id}>
          <NewsListItem id={news.id} title={news.title} abstract={news.abstract} section={news.section} byline={news.byline} media={news.media} publishedDate={news.published_date} alt={!(cards++ % 2)}/>
        </Col>
      )}
    </Row>;
  }
}

NewsListPage.propTypes = {
  newsList: PropTypes.array,
  newsLoading: PropTypes.bool,
  newsMessage: PropTypes.string,
  newsStatus: PropTypes.bool,
  fetchNews: PropTypes.func,
}

const mapStateToProps = state => {
  return {
    newsList: selectNewsList(state),
    newsLoading: selectNewsLoading(state),
    newsMessage: selectNewsMessage(state),
    newsStatus: selectNewsStatus(state),
  };
};

export function mapDispatchToProps(dispatch) {
  return {
    fetchNews: () => dispatch(fetchNews())
  };
}

const withConnect = connect(
  mapStateToProps,
  mapDispatchToProps,
);

export default compose(
  withConnect,
  memo,
)(NewsListPage);
