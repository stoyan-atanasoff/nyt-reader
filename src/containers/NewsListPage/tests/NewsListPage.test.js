/**
 * Test the NewsListPage
 */

import React from 'react';
import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import { Router } from 'react-router-dom';
import configureStore from 'redux-mock-store'
import { fromJS } from "immutable";
import NewsListPage from '../index';
import { history } from '../../../store/configureStore';

const initialState = {
  news: fromJS({
    newsList: [],
    newsLoading: false,
    newsStatus: false,
    newsMessage: '',
  })
};
const mockStore = configureStore();
let store;

describe('<NewsListPage />', () => {

  beforeAll(() => {
    store = mockStore(initialState)
  });

  it('News loading should render and match the snapshot', () => {
    store = mockStore({
      news: fromJS({
        newsList: [],
        newsLoading: true,
        newsStatus: false,
        newsMessage: '',
      })
    });

    const {
      container: { firstChild },
    } = render(
      <Provider store={store}>
        <NewsListPage/>
      </Provider>
    );
    expect(firstChild).toMatchSnapshot();
  });

  it('News loaded successfully should render correctly', async () => {

    const newsItem = [{
      "url": "https://www.nytimes.com/2018/06/04/us/graduate-quotes-trump-obama.html",
      "adx_keywords": "Commencement Speeches;Bowling, Ben;Trump, Donald J;Obama, Barack;Kentucky;United States Politics and Government",
      "column": null,
      "section": "U.S.",
      "byline": "By LAURA M. HOLSON",
      "type": "Article",
      "title": "Kentucky Crowd Cheers Valedictorian’s Trump Quote, Then Learns Obama Said It",
      "abstract": "A high school valedictorian quoted Trump. The crowd roared. Then: “Just kidding,” he said. “That was Barack Obama.”",
      "published_date": "2018-06-04",
      "source": "The New York Times",
      "id": 100000005935794,
      "asset_id": 100000005935794,
      "views": 1,
      "des_facet": ["COMMENCEMENT SPEECHES"],
      "org_facet": ["UNITED STATES POLITICS AND GOVERNMENT"],
      "per_facet": ["BOWLING, BEN", "TRUMP, DONALD J", "OBAMA, BARACK"],
      "geo_facet": ["KENTUCKY"],
      "media": [{
        "type": "image",
        "subtype": "photo",
        "caption": "Ben Bowling, the valedictorian of Bell County High School.",
        "copyright": "via Facebook",
        "approved_for_syndication": 0,
        "media-metadata": [{
          "url": "https://static01.nyt.com/images/2018/06/04/us/05xp-graduation2/05xp-graduation2-thumbStandard.jpg",
          "format": "Standard Thumbnail",
          "height": 75,
          "width": 75
        }, {
          "url": "https://static01.nyt.com/images/2018/06/04/us/05xp-graduation2/05xp-graduation2-mediumThreeByTwo210.jpg",
          "format": "mediumThreeByTwo210",
          "height": 140,
          "width": 210
        }, {
          "url": "https://static01.nyt.com/images/2018/06/04/us/05xp-graduation2/05xp-graduation2-mediumThreeByTwo440.jpg",
          "format": "mediumThreeByTwo440",
          "height": 293,
          "width": 440
        }]
      }],
      "uri": "nyt://article/02160478-cbb7-538d-87fa-a5777305e833"
    }];
    store = mockStore({
      news: fromJS({
        newsList: newsItem,
        newsLoading: false,
        newsStatus: true,
        newsMessage: '',
      })
    });
    const {
      container: { firstChild },
    } = render(
      <Provider store={store}>
        <Router history={history}>
          <NewsListPage/>
        </Router>
      </Provider>
    );

    expect(firstChild).toMatchSnapshot();
  });

  it('News load failed render and match the snapshot', () => {
    store = mockStore({
      news: fromJS({
        newsList: [],
        newsLoading: false,
        newsStatus: false,
        newsMessage: 'There is an error while loading news',
      })
    });

    const {
      container: { firstChild },
    } = render(
      <Provider store={store}>
        <NewsListPage/>
      </Provider>
    );
    expect(firstChild).toMatchSnapshot();
  });

});
