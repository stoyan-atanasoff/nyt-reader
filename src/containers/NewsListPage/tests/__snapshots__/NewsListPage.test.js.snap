// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NewsListPage /> News load failed render and match the snapshot 1`] = `
<div
  class="ant-row-flex ant-row-flex-center news-message"
  lign="top"
>
  <div
    class="ant-col"
  >
    <div
      class="ant-alert ant-alert-error ant-alert-no-icon"
      data-show="true"
    >
      <span
        class="ant-alert-message"
      >
        There is an error while loading news
      </span>
      <span
        class="ant-alert-description"
      />
    </div>
  </div>
</div>
`;

exports[`<NewsListPage /> News loaded successfully should render correctly 1`] = `
<div
  class="ant-row-flex"
  lign="top"
>
  <div
    class="ant-col ant-col-xs-24 ant-col-sm-24"
  >
    <div
      class="blog-card "
    >
      <div
        class="meta"
      >
        <div
          class="photo"
          style="background-image: url(https://static01.nyt.com/images/2018/06/04/us/05xp-graduation2/05xp-graduation2-mediumThreeByTwo440.jpg);"
        />
        <ul
          class="details"
        >
          <li
            class="author"
          >
            <svg
              aria-hidden="true"
              class="svg-inline--fa fa-user fa-w-14 "
              data-icon="user"
              data-prefix="fas"
              focusable="false"
              role="img"
              viewBox="0 0 448 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M224 256c70.7 0 128-57.3 128-128S294.7 0 224 0 96 57.3 96 128s57.3 128 128 128zm89.6 32h-16.7c-22.2 10.2-46.9 16-72.9 16s-50.6-5.8-72.9-16h-16.7C60.2 288 0 348.2 0 422.4V464c0 26.5 21.5 48 48 48h352c26.5 0 48-21.5 48-48v-41.6c0-74.2-60.2-134.4-134.4-134.4z"
                fill="currentColor"
              />
            </svg>
             
            By LAURA M. HOLSON
          </li>
          <li
            class="date"
          >
            <svg
              aria-hidden="true"
              class="svg-inline--fa fa-calendar fa-w-14 "
              data-icon="calendar"
              data-prefix="fas"
              focusable="false"
              role="img"
              viewBox="0 0 448 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M12 192h424c6.6 0 12 5.4 12 12v260c0 26.5-21.5 48-48 48H48c-26.5 0-48-21.5-48-48V204c0-6.6 5.4-12 12-12zm436-44v-36c0-26.5-21.5-48-48-48h-48V12c0-6.6-5.4-12-12-12h-40c-6.6 0-12 5.4-12 12v52H160V12c0-6.6-5.4-12-12-12h-40c-6.6 0-12 5.4-12 12v52H48C21.5 64 0 85.5 0 112v36c0 6.6 5.4 12 12 12h424c6.6 0 12-5.4 12-12z"
                fill="currentColor"
              />
            </svg>
             
            2018-06-04
          </li>
        </ul>
      </div>
      <div
        class="description"
      >
        <h1>
          Kentucky Crowd Cheers Valedictorian’s Trump Quote, Then Learns Obama Said It
        </h1>
        <h2>
          U.S.
        </h2>
        <p>
          A high school valedictorian quoted Trump. The crowd roared. Then: “Just kidding,” he said. “That was Barack Obama.”
        </p>
        <p
          class="read-more"
        >
          <a
            href="/news/100000005935794"
          >
            Read More 
            <svg
              aria-hidden="true"
              class="svg-inline--fa fa-arrow-right fa-w-14 "
              data-icon="arrow-right"
              data-prefix="fas"
              focusable="false"
              role="img"
              viewBox="0 0 448 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M190.5 66.9l22.2-22.2c9.4-9.4 24.6-9.4 33.9 0L441 239c9.4 9.4 9.4 24.6 0 33.9L246.6 467.3c-9.4 9.4-24.6 9.4-33.9 0l-22.2-22.2c-9.5-9.5-9.3-25 .4-34.3L311.4 296H24c-13.3 0-24-10.7-24-24v-32c0-13.3 10.7-24 24-24h287.4L190.9 101.2c-9.8-9.3-10-24.8-.4-34.3z"
                fill="currentColor"
              />
            </svg>
          </a>
        </p>
      </div>
    </div>
  </div>
</div>
`;

exports[`<NewsListPage /> News loading should render and match the snapshot 1`] = `
<div
  class="loader"
>
  <div
    class="ant-spin ant-spin-spinning"
  >
    <span
      class="ant-spin-dot ant-spin-dot-spin"
    >
      <i
        class="ant-spin-dot-item"
      />
      <i
        class="ant-spin-dot-item"
      />
      <i
        class="ant-spin-dot-item"
      />
      <i
        class="ant-spin-dot-item"
      />
    </span>
  </div>
  <br />
  
</div>
`;
